### ISSUE TITLE
Any reason Halide is working differently in Debug or Release builds?

### ISSUE THREAD

**Issue by mcourteaux** [halide/Halide#4378]
I am seeing the strangest behaviour. I narrowed it down; I'm sure that my problem is on Halide's part.
I create a source buffer  and a destination buffer and pass them into an AOT function. The curious thing is that if I compile my project in Release mode, Halide seems to work correctly; whereas in Debug mode, it fails to produce correct results.

Things I've checked:
 - The input is equal in both cases.
 - The output is incorrect in Debug.
 - The output is correct in Release (RelWithDebInfo, for cmake).
 - The generated AOT static libraries which I link with are bit-identical in a Debug or Release build.

This makes me believe the issue is could be due to:
 - Incorrect calling conventions?
 - A bug in the Halide Runtime headers used to create, read, and write buffers?

Any other reasons why this could happen? Any hints where to look?

**Comment by steven-johnson** [comment]
(1) When you say "Debug" I presume you mean "Halide itself is compiled with the C++ compiler using Debug flags rather than Release+Optimized Flags"? (As opposed to "using the `debug` feature in my Target flags.)
(2) What specific compilers/platforms/etc are you actually using?

**Comment by abadams** [comment]
The Halide runtime is part of the AOT static library, so if it's bit-identical it's hard to imagine there's a difference, and the generated headers are pretty minimal, and only contain struct definitions and function signatures.

The -debug target flag might help you diagnose what's going on - it prints all the input/output buffer metadata for you to sanity check.

**Comment by mcourteaux** [comment]
@steven-johnson (1) No, I'm compiling my own project with Debug or Release. I downloaded Halide from the releases tab here on GitHub. So I'm linking my own in-debug-compiled program with the generated static library from the AOT generator.

@steven-johnson (2):
```
❯ clang++ --version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin19.0.0
Thread model: posix
```

@abadams I was thinking about the `#include <HalideBuffer.h>`, which I use to construct the input and output buffers.  Concerning the "only contain struct definitions"; I'm thinking about different struct field padding / alignment depending on the compiler optimisation level? I tried the debug flag, and for a working Release build, I got:
```
Entering Pipeline neonraw_demosaic_bilinear_RGGB
Target: x86-64-osx-avx-avx2-debug-f16c-fma-sse41
 Input int32 offset_x: 0
 Input int32 offset_y: 0
 Input uint16 bit_shift: 1
 Input Buffer black_levels: buffer(0, 0x0, 0x7000096f7a10, 0, uint16, {0, 4, 1})
 Input Buffer raw_buffer: buffer(0, 0x0, 0x11a7f0000, 0, uint16, {0, 6032, 1}, {0, 4032, 6032})
 Output Buffer output: buffer(0, 0x0, 0x11dda7080, 0, uint16, {0, 6032, 1}, {0, 4032, 6032}, {0, 3, 24321024})
Exiting Pipeline neonraw_demosaic_bilinear_RGGB
```
For an erroneous Debug build, I got:
```
Entering Pipeline neonraw_demosaic_bilinear_RGGB
Target: x86-64-osx-avx-avx2-debug-f16c-fma-sse41
 Input int32 offset_x: 0
 Input int32 offset_y: 0
 Input uint16 bit_shift: 1
 Input Buffer black_levels: buffer(0, 0x0, 0x700007531ed0, 0, uint16, {0, 4, 1})
 Input Buffer raw_buffer: buffer(0, 0x0, 0x125b81800, 0, uint16, {0, 6032, 1}, {0, 4032, 6032})
 Output Buffer output: buffer(0, 0x0, 0x12b4f3880, 0, uint16, {0, 6032, 1}, {0, 4032, 6032}, {0, 3, 24321024})
Exiting Pipeline neonraw_demosaic_bilinear_RGGB
```
I don't know exactly what all these numbers are, but they at least seem reasonable, and identical, except for the memory addresses of course.

As extra info, I'm using Halide to generate a bilinear demosaicer (from a RAW photo file): The output when it goes wrong looks like this:
<img width="1360" alt="Screenshot 2019-11-05 at 23 07 45" src="https://user-images.githubusercontent.com/845012/68250503-5482cb00-0021-11ea-85cb-ce1186a1e96c.png">
Release mode gives perfect results; every time. One can cleary see some diagonal patterns, as if the data has the wrong stride; or is transposed? Or the dimension sizes are swapped? There is *some* structure in there... I did some other sanity checks to confirm the input buffers; these seem to be fine.

So to recap, my gut feeling goes to:
 - calling conventions,
 - struct padding,
 - struct alignment assumptions,
 - a bug in `HalideBuffer.h`, incorrectly acting as "wrapper" for `halide_buffer_t`,
 - some severe incompatibility between Halide and Address Sanitiser (as I'm using ASAN only in Debug builds).

**Comment by abadams** [comment]
Interesting, that really does look like a stride issue, but the debug output does indeed look totally reasonable. The numbers are two gpu-related fields, then the base memory address, then the dirty flags (only relevant for GPU), then the type, then min/extent/stride in each dimension.

One thing you could try to rule in or out HalideBuffer.h is just constructing halide_buffer_t objects directly and passing those in instead of using Halide::Runtime::Buffer.

Turning off ASAN would also be a good test.

**Comment by mcourteaux** [comment]
I turned of ASAN: no luck.
I converted my code to plain `halide_buffer_t` usage: exactly the same results: Debug wrong, Release fine.
As I'm getting more and more suspicious, I will do some more paranoid checks on my input tomorrow. I'll try to SHA-1 my input buffer. Thanks for the help so far!

**Comment by abadams** [comment]
That leaves calling convention weirdness and struct weirdness. But the debug mode dumps the struct contents as understood by Halide, and it's the same in both cases, so that can't be it. Halide is understanding the arguments in the same way in both cases.

So surely it has to be a difference in the contents of the input buffers? I'd agree that hashing or spot-checking the buffer contents is the next step.

**Comment by mcourteaux** [comment]
Okay, I implemented SHA-1 checking, and it seems to have different inputs in Debug and Release. My prior checks was only averaging out the first 1000 values of the input, which is less then one row of data. The fault is now on either my side, or RawSpeeds side. Thank you for your assistance!

**Comment by mcourteaux** [comment]
I've been mistaken. The input is the same...

**Comment by mcourteaux** [comment]
To be completely open about the details: RawSpeed gives loads a RAW file into a buffer, but the actual image is padded into the buffer (both padded horizontally, and vertically). The center region containing the actual photo data hashes exactly the same in Debug and Release. The complete buffer with padding around is different. But this shouldn't be important, as my Halide function definition takes this into account. Will do more investigations on this...

**Comment by mcourteaux** [comment]
I found it! The RawSpeed `uncropped_dim.x*sizeof(uint16_t)` is apparently not equivalent to the `pitch`. In Release, `pitch` and `uncropped_dim.x*sizeof(uint16_t)` were the same, and in debug they are different. One needs to use `pitch / sizeof(uint16_t)` instead to assign as width of the Halide buffer.
